src/Pos/Util/Modifier.hs

```diff
@@ -0,0 +1,142 @@
+{-# LANGUAGE TypeFamilies #-}
+
+-- | Wrapper for modifier pattern which is used intensively in code
+
+module Pos.Util.Modifier
+       ( MapModifier
+       , lookupM
+       , lookup
+       , keysM
+       , keys
+       , valuesM
+       , values
+       , toListM
+       , toList
+       , insertions
+       , deletions
+
+       , insert
+       , delete
+
+       , mapMaybeM
+       ) where
+
+import           Data.Hashable       (Hashable)
+import qualified Data.HashMap.Strict as HM
+import           Universum           hiding (toList)
+
+-- | 'MapModifier' is a type which collects modifications (insertions
+-- and deletions) of something map-like.
+newtype MapModifier k v = MapModifier
+    { getMapModifier :: HashMap k (Maybe v)
+    } deriving (Show)
+
+instance (Eq k, Hashable k) =>
+         Monoid (MapModifier k v) where
+    mempty = MapModifier mempty
+    mappend m1 (MapModifier m2) = HM.foldlWithKey' step m1 m2
+      where
+        step m k Nothing  = delete k m
+        step m k (Just v) = insert k v m
+
+instance (Eq k, Hashable k) => Semigroup (MapModifier k v)
+
+-- | Perform monadic lookup taking 'MapModifier' into account.
+lookupM
+    :: (Applicative m, Eq k, Hashable k)
+    => (k -> m (Maybe v)) -> k -> MapModifier k v -> m (Maybe v)
+lookupM getter k (MapModifier m) =
+    case HM.lookup k m of
+        Nothing       -> getter k
+        Just Nothing  -> pure Nothing
+        Just (Just x) -> purer x
+
+-- | Perform lookup taking 'MapModifier' into account.
+lookup
+    :: (Eq k, Hashable k)
+    => (k -> Maybe v) -> k -> MapModifier k v -> Maybe v
+lookup getter k = runIdentity . lookupM (Identity . getter) k
+
+-- | Get keys of something map-like in Functor context taking
+-- 'MapModifier' into account.
+keysM
+    :: (Functor m, Eq k, Hashable k)
```

Comment:
Suppose we have something map-like, e. g. key-value DB. There are a lot of key-value pairs there. Keys are small, values are big. We also have a modifier. We want to get all keys taking modifier into account. `m [k]` is an action which produces all keys of this map-like thing.